File: kb-rag-web/src/api/chatDiagnostics.ts
/** 预览阶段实测值；缺失和 null 均表示未采集，零是有效测量。 */
export interface ChatDiagnostics {
  request_id?: string;
  outcome: 'SUCCEEDED' | 'FAILED' | 'CANCELLED';
  failed_stage?: 'CONFIGURATION' | 'RETRIEVAL' | 'GENERATION' | null;
  configuration_ms?: number | null;
  retrieval_ms?: number | null;
  generation_ms?: number | null;
  first_delta_ms?: number | null;
  total_ms: number;
}

const durationFields = ['configuration_ms', 'retrieval_ms', 'generation_ms', 'first_delta_ms', 'total_ms'] as const;

/** 可选诊断异常不破坏正文协议，也不把非法值变成零耗时。 */
export function readChatDiagnostics(value: unknown): ChatDiagnostics | undefined {
  if (!value || typeof value !== 'object' || Array.isArray(value)) return undefined;
  const data = value as Record<string, unknown>;
  if (typeof data.outcome !== 'string' || !['SUCCEEDED', 'FAILED', 'CANCELLED'].includes(data.outcome)) return undefined;
  if (typeof data.total_ms !== 'number') return undefined;
  if (data.failed_stage != null && (typeof data.failed_stage !== 'string'
    || !['CONFIGURATION', 'RETRIEVAL', 'GENERATION'].includes(data.failed_stage))) return undefined;
  for (const field of durationFields) {
    const duration = data[field];
    if (duration != null && (typeof duration !== 'number' || !Number.isSafeInteger(duration) || duration < 0)) return undefined;
  }
  return {
    request_id: typeof data.request_id === 'string' ? data.request_id : undefined,
    outcome: data.outcome as ChatDiagnostics['outcome'],
    failed_stage: data.failed_stage as ChatDiagnostics['failed_stage'],
    configuration_ms: data.configuration_ms as number | null | undefined,
    retrieval_ms: data.retrieval_ms as number | null | undefined,
    generation_ms: data.generation_ms as number | null | undefined,
    first_delta_ms: data.first_delta_ms as number | null | undefined,
    total_ms: data.total_ms,
  };
}

File: kb-rag-server/kb-app/src/main/java/io/kbrag/app/openapi/ChatDiagnostics.java
package io.kbrag.app.openapi;

import io.kbrag.app.retrieval.RerankTiming;

/** 预览请求的单调时钟测量；未执行的阶段和未收到的首段保持为空。 */
public record ChatDiagnostics(Outcome outcome, Stage failedStage, Long configurationMs,
                              Long retrievalMs, Long generationMs, Long firstDeltaMs, long totalMs,
                              RerankTiming rerank) {

    /** 兼容未单独采集重排的调用，缺失的子阶段保持为空。 */
    public ChatDiagnostics(Outcome outcome, Stage failedStage, Long configurationMs,
                           Long retrievalMs, Long generationMs, Long firstDeltaMs, long totalMs) {
        this(outcome, failedStage, configurationMs, retrievalMs, generationMs, firstDeltaMs, totalMs, null);
    }

    /** 编排阶段；重排子段包含在 RETRIEVAL 总段内，不能再次相加。 */
    public enum Stage { CONFIGURATION, RETRIEVAL, GENERATION }

    /** 请求最终结果，不以收到诊断事件代表回答成功。 */
    public enum Outcome { SUCCEEDED, FAILED, CANCELLED }
}
